File: src/main/java/de/jj22/telegram_bot_rpg/Game/Entities/GameDatabase.java
package de.jj22.telegram_bot_rpg.Game.Entities;

import java.sql.Connection;
import java.sql.DriverManager;
import java.sql.PreparedStatement;
import java.sql.ResultSet;
import java.sql.SQLException;
import java.sql.Statement;

public class GameDatabase {
	final private Connection connection;

	public GameDatabase(String filename) throws SQLException {
		connection = DriverManager.getConnection("jdbc:sqlite:" + filename);
		_initDatabase();
	}

	private void _initDatabase() throws SQLException {
		var asd = connection.createStatement();
		asd.execute("CREATE TABLE IF NOT EXISTS FirstEntity ( entity_id INTEGER PRIMARY KEY)");
		asd.execute(
				"CREATE TABLE IF NOT EXISTS Entity ( entity_id INTEGER PRIMARY KEY AUTOINCREMENT, name TEXT, description TEXT )");
		asd.execute(
				"CREATE TABLE IF NOT EXISTS EntityList ( list_id INTEGER , list_index INTEGER, linked_entity_id INTEGER, PRIMARY KEY(list_id, list_index) )");
		asd.execute(
				"CREATE TABLE IF NOT EXISTS Player ( player_id INTEGER PRIMARY KEY, entity_id INTEGER, timestamp DATETIME DEFAULT CURRENT_TIMESTAMP )");
		asd.close();
	}

	public void deleteEntity(int entity_id) throws SQLException {
		preparedStatement("DELETE FROM Entity WHERE entity_id=?", entity_id).executeUpdate();
	}

	public void deleteEntityList(int list_id) throws SQLException {
		preparedStatement("DELETE FROM EntityList WHERE list_id=?", list_id).executeUpdate();
	}

	public void deleteFirstEntity() throws SQLException {
		connection.createStatement().execute("DELETE FROM FirstEntity");
	}

	public void deletePlayer(int player_id) throws SQLException {
		preparedStatement("DELETE FROM Player WHERE player_id=?", player_id).executeUpdate();
	}

	public ResultSet insertEntity(String name, String description) throws SQLException {
		final var stmnt = preparedStatement("INSERT INTO Entity (name, description) VALUES (?, ?)", name, description);
		stmnt.executeUpdate();
		return stmnt.getGeneratedKeys();
	}

	public void insertEntityList(int list_id, int index, int entity_id) throws SQLException {
		preparedStatement("INSERT INTO EntityList VALUES (?, ?, ?)", list_id, index, entity_id).executeUpdate();
	}

	public void insertPlayer(int player_id, int entity_id) throws SQLException {
		preparedStatement("INSERT INTO Player (player_id, entity_id) VALUES (?, ?)", player_id, entity_id).executeUpdate();
	}

	private PreparedStatement preparedStatement(String pre, Object... vals) throws SQLException {
		final var statement = connection.prepareStatement(pre, Statement.RETURN_GENERATED_KEYS);

		for (int i = 0; i < vals.length; i++) {
			statement.setObject(i + 1, vals[i]);
		}

		return statement;
	}

	/**
	 * 
	 * @param entity_id
	 * @return Integer entity_id, String name, String description
	 * @throws SQLException
	 */
	public ResultSet selectEntity(int entity_id) throws SQLException {
		final var statement = preparedStatement("SELECT entity_id, name, description FROM Entity WHERE entity_id=?", entity_id);
		return statement.executeQuery();
	}

	/**
	 * 
	 * @param list_id
	 * @param index
	 * @return Integer list_id, Integer list_index , Integer linked_entity_id
	 * @throws SQLException
	 */
	public ResultSet selectEntityList(int list_id, int index) throws SQLException {
		final var statement = preparedStatement("SELECT list_id, list_index, linked_entity_id FROM EntityList WHERE (list_id=? AND list_index=?)", list_id,
				index);
		return statement.executeQuery();
	}

	/**
	 * 
	 * @return Integer entity_id
	 * @throws SQLException
	 */
	public ResultSet selectFirstEntity() throws SQLException {
		final var statement = preparedStatement("SELECT entity_id FROM FirstEntity");
		return statement.executeQuery();
	}

	/**
	 * 
	 * @param player_id
	 * @return Integer player_id, Integer entity_id
	 * @throws SQLException
	 */
	public ResultSet selectPlayer(int player_id) throws SQLException {
		final var statement = preparedStatement("SELECT player_id, entity_id FROM Player WHERE player_id=?", player_id);
		return statement.executeQuery();
	}

	public void setFirstEntity(int entity_id) throws SQLException {
		deleteFirstEntity();
		preparedStatement("INSERT INTO FirstEntity VALUES (?)", entity_id).execute();
	}

	public void updateEntityDescription(int entity_id, String description) throws SQLException {
		preparedStatement("UPDATE Entity SET description=? WHERE entity_id=?", description, entity_id).execute();
	}

	public void updateEntityList(int list_id, int index, int entity_id) throws SQLException {
		preparedStatement("UPDATE EntityList SET entity_id=? WHERE (list_id=? AND list_index=?)", entity_id, list_id,
				index).execute();
	}

	public void updateEntityName(int entity_id, String name) throws SQLException {
		preparedStatement("UPDATE Entity SET name=? WHERE entity_id=?", name, entity_id).execute();
	}

	public void updatePlayerTimestamp(int player_id) throws SQLException {
		preparedStatement("UPDATE Player SET timestamp=CURRENT_TIMESTAMP WHERE player_id=?", player_id).execute();
	}

}

File: Game/EntitySystem/GameDatabase.ts
import SQLWrapper from "../../utils/SQLWrapper";

export default class GameDatabase {
	readonly #sql: SQLWrapper;

	constructor(fileName: string) {
		this.#sql = new SQLWrapper(fileName);
		this.prepareDB();
	}

	private async prepareDB() {
		await this.#sql.query("CREATE TABLE IF NOT EXISTS Entity (entity_id INT PRIMARY KEY, name TEXT, description TEXT);");
		await this.#sql.query("CREATE TABLE IF NOT EXISTS LinkedEntities (entity_id INT, i INT, linked_entity_id INT);");

		await this.#sql.query("CREATE TABLE IF NOT EXISTS Player (player_id INT PRIMARY KEY, platform_id TEXT, platform TEXT, entity_id INT);");
		await this.#sql.query("CREATE TABLE IF NOT EXISTS Role (role_name TEXT PRIMARY KEY);");
		// await this.sql.query("INSERT INTO Role VALUES ('admin');");
		// await this.sql.query("INSERT INTO Role VALUES ('player');");
		await this.#sql.query("CREATE TABLE IF NOT EXISTS PlayerRoles (id INT, role_name TEXT);");
		// await this.sql.query("INSERT INTO Player VALUES (1,'1','console',NULL);");
	}


	async selectEntity(entity_id: number) {
		return this.#sql.query("SELECT * FROM Entity WHERE entity_id=?", entity_id);
	}

	async selectEntityIdWherePlayer(platform_id: string, platform: string) {
		return this.#sql.query("SELECT entity_id FROM Player WHERE platform_id=? AND platform=?", platform_id, platform);
	}

	async selectPlayerRoles(platform_id: string, platform: string) {
		const res = await this.#sql.query("SELECT role_name FROM PlayerRoles NATURAL JOIN Player WHERE platform_id=?, platform=?", platform_id, platform);
		return res[0];
	}

	async insertPlayer(platform_id: string, platform: string, entity_id: number) {
		const res = await this.#sql.query("INSERT INTO Player (platform_id, platform, entity_id) VALUES (?,?,?)", platform_id, platform, entity_id);
		console.debug("GameDatabase.insertPlayer() ", res);
		return res;

	}

	async updatePlayerEntityID(entity_id: number) {
		const res = await this.#sql.query("UPDATE Player (entity_id) VALUES (?)", entity_id);
		console.debug("GameDatabase.updatePlayerEntityID() ", res);
		return res;
	}

	async insertNewEntity(name: string, description: string) {
		const res = await this.#sql.query("INSERT INTO Entity (name, description) VALUES (?,?)", name, description);
		console.debug("GameDatabase.insertNewEntity() ", res);
		return res;
	}
	async selectPlayer(platform_id: string, platform: string) {
		return this.#sql.query("SELECT * FROM Player WHERE platform_id=? AND platform=?", platform_id, platform);
	}
}